lib/spack/docs/packaging_guide.rst

```diff
@@ -5388,6 +5388,86 @@ follows:
 
 .. _package-lifecycle:
 
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Add detection tests to packages
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To ensure that a software is detected correctly for multiple configurations
+and on different systems users can write a ``detection_test.yaml`` file and
+put it in the same directory as the corresponding ``package.py`` file.
+This YAML file contains enough information for Spack to mock an environment
+and try to check if the detection logic yields the results that are expected.
+
+As a general rule, attributes at the top-level of ``detection_test.yaml``
+represent search mechanisms and they all map to a list of tests that should confirm
+the validity of each package detection logic.
+
+""""""""""""""""""""""""""
+Tests for PATH inspections
+""""""""""""""""""""""""""
+
+Detection tests insisting on ``PATH`` inspections are listed under
+the ``paths`` attribute:
+
+.. code-block:: yaml
+
+   paths:
+   - layout:
+     - subdir: [bin]
+       name: clang-3.9
+       output: |
+         echo "clang version 3.9.1-19ubuntu1 (tags/RELEASE_391/rc2)"
+         echo "Target: x86_64-pc-linux-gnu"
+         echo "Thread model: posix"
+         echo "InstalledDir: /usr/bin"
+     - subdir: [bin]
+       name: clang++-3.9
+       output: |
+         echo "clang version 3.9.1-19ubuntu1 (tags/RELEASE_391/rc2)"
+         echo "Target: x86_64-pc-linux-gnu"
+         echo "Thread model: posix"
+         echo "InstalledDir: /usr/bin"
+     results:
+     - spec: 'llvm@3.9.1 +clang~lld~lldb'
```

Comment:
I'm worried that this format is very obtuse. Would we lose any information if we switched to something like:

```suggestion
   - layout:
     - executable: bin/clang-3.9
       output: |
        "clang version 3.9.1-19ubuntu1 (tags/RELEASE_391/rc2)"
        "Target: x86_64-pc-linux-gnu"
        "Thread model: posix"
        "InstalledDir: /usr/bin"
     - executable: bin/clang++-3.9
       output: |
        "clang version 3.9.1-19ubuntu1 (tags/RELEASE_391/rc2)"
        "Target: x86_64-pc-linux-gnu"
        "Thread model: posix"
        "InstalledDir: /usr/bin"
     specs:
     - 'llvm@3.9.1 +clang~lld~lldb'
```

This would seem to remove unnecessary verbosity.

Reply:
Done in https://github.com/spack/spack/pull/18175/commits/d40f73f6d71e9a1583411d306d957a264f50a05b